controller/src/main/java/org/jboss/as/controller/CapabilityRegistry.java

```diff
@@ -0,0 +1,745 @@
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Registers an additional requirement a capability has beyond what it was aware of when {@code capability}
+     * was passed to {@link #registerCapability(RuntimeCapabilityRegistration)}. Used for cases
+     * where a capability optionally depends on another capability, and whether or not that requirement is needed is
+     * not known when the capability is first registered.
+     *
+     * @param requirement the requirement
+     * @throws java.lang.IllegalArgumentException if no matching capability is currently
+     *                                            {@link #registerCapability(RuntimeCapabilityRegistration) registered} for either {@code required} or {@code dependent}
+     */
+    @Override
+    public void registerAdditionalCapabilityRequirement(RuntimeRequirementRegistration requirement) {
+        writeLock.lock();
+        try {
+            registerRequirement(requirement);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    private void registerRequirement(RuntimeRequirementRegistration requirement) {
+        try {
+            writeLock.lock();
+            CapabilityId dependentId = requirement.getDependentId();
+            if (!capabilities.containsKey(dependentId)) {
+                throw ControllerLogger.MGMT_OP_LOGGER.unknownCapabilityInContext(dependentId.getName(),
+                        dependentId.getContext().getName());
+            }
+            Map<CapabilityId, Map<String, RuntimeRequirementRegistration>> requirementMap =
+                    requirement.isRuntimeOnly() ? runtimeOnlyRequirements : requirements;
+
+            Map<String, RuntimeRequirementRegistration> dependents = requirementMap.get(dependentId);
+            if (dependents == null) {
+                dependents = new HashMap<>();
+                requirementMap.put(dependentId, dependents);
+            }
+            RuntimeRequirementRegistration existing = dependents.get(requirement.getRequiredName());
+            if (existing == null) {
+                dependents.put(requirement.getRequiredName(), requirement);
+            } else {
+                existing.addRegistrationPoint(requirement.getOldestRegistrationPoint());
+            }
+            modified = true;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Remove a previously registered requirement for a capability.
+     *
+     * @param requirementRegistration the requirement. Cannot be {@code null}
+     * @see #registerAdditionalCapabilityRequirement(org.jboss.as.controller.capability.registry.RuntimeRequirementRegistration)
+     */
+    @Override
+    public void removeCapabilityRequirement(RuntimeRequirementRegistration requirementRegistration) {
+        // We don't know if this got registered as an runtime-only requirement or a hard one
+        // so clean it from both maps
+        try {
+            writeLock.lock();
```

Comment:
All the lock acquisitions need to be moved outside the try/finally that releases them. See Java Concurrency in Practice listing 13.2, - 13.5.